### PR TITLE
Refactor taint_manager to do not use getPod and getNode stubs

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR refactors taint_manager to eliminate the getPod and getNode stubs. It will allow easier adding or unit tests. 
It is a preparatory code change before 
[Append new pod conditions when deleting pods to indicate the reason for pod deletion #110959](https://github.com/kubernetes/kubernetes/pull/110959).

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP-3329: Retriable and non-retriable Pod failures for Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures)

### CODE REVIEW

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go [hunk 1]
```diff
@@ -219,10 +178,14 @@ func TestCreatePod(t *testing.T) {
...
+		nodeIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil)
+		controller := setupNewNoExecuteTaintManager(ctx, fakeClientset, podIndexer, nodeIndexer)
 		controller.recorder = testutil.NewFakeRecorder()
 		go controller.Run(ctx)
 		controller.taintedNodes = item.taintedNodes
+
+		podIndexer.Add(item.pod)
 		controller.PodUpdated(nil, item.pod)
 		// wait a bit
 		time.Sleep(timeForControllerToProgress)
```
Comment (alculquicondor) [inline pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:191]: For a follow up (not required for your enhancement): replace this with `wait.Until`

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go [hunk 2]
```diff
@@ -133,6 +88,10 @@ func addTaintsToNode(node *v1.Node, key, value string, indices []int) *v1.Node {
 	return node
 }
 
+func setupNewNoExecuteTaintManager(ctx context.Context, fakeClientSet *fake.Clientset, podIndexer cache.Indexer, nodeIndexer cache.Indexer) *NoExecuteTaintManager {
+	return NewNoExecuteTaintManager(ctx, fakeClientSet, corelisters.NewPodLister(podIndexer), corelisters.NewNodeLister(nodeIndexer), getPodsAssignedToNode(fakeClientSet))
```
Comment (alculquicondor) [inline pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:92]: You can create the indexers from the clientset using an informer factory:

```golang
informerFactory := informers.SharedInformerFactory(fakeClientset, 0)
return NewNoExecuteTaintManager(ctx, fakeClientset, informerFactory.Core().V1().Pods(), informerFactory.Core().V1().Nodes(), getPodsAssignedToNode(fakeClientSet))
```

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go [hunk 3]
```diff
@@ -219,10 +178,14 @@ func TestCreatePod(t *testing.T) {
...
-		controller := NewNoExecuteTaintManager(ctx, fakeClientset, (&podHolder{pod: item.pod}).getPod, getNodeFromClientset(fakeClientset), getPodsAssignedToNode(fakeClientset))
+		podIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil)
+		nodeIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil)
+		controller := setupNewNoExecuteTaintManager(ctx, fakeClientset, podIndexer, nodeIndexer)
 		controller.recorder = testutil.NewFakeRecorder()
 		go controller.Run(ctx)
 		controller.taintedNodes = item.taintedNodes
+
+		podIndexer.Add(item.pod)
 		controller.PodUpdated(nil, item.pod)
```
Comment (alculquicondor) [inline pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:189]: I think that you don't need to explicitly call this function if you already added the Pod to the indexer (or the clientset once you do the change I suggested above). The event handler should take care of it.

Comment (mimowo) [reply pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:191]: I now use indexers based on the fake clientSet, but still have to call this Add/Update functions. It seems that job_controller_test.go follows this pattern too.

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**